### ISSUE TITLE
How can I trigger a socket acknowledgment?

### ISSUE THREAD

**Issue by Sanuja91** [Links2004/arduinoWebSockets#457]
How can I trigger a socket acknowledgment? Basically a callback function is sent in the socket through socket io from a node js server.. How can I trigger this?

**Comment by Links2004** [comment]
this is done via the acknowledgment (ACK) handling.
for a more detailed view about how the socket.io protocol is working on a JSON / data level I recommend taking a look at:
https://github.com/socketio/socket.io-protocol

**Comment by Sanuja91** [comment]
Thanks for the quick response! Could please implement a public API for sIOtype_ACK message.. That would be amazing! Additionally, kindly show a coded implementation in the examples. Personally, this is the only hiccup I found in this library.. Great job with this library.. Really saved my ass!

**Comment by Sanuja91** [comment]
Hi Links, will you be implementing this? Till then can you kindly show how to call the callback function from the Arduino?

**Comment by anselmpaul** [comment]
With the earlier SocketIO implementation, you could acknowledge a new connection by sending `webSocket.sendTXT("5");` and had to send a heartbeat Message every <30 seconds like this
`if(isConnected) {
        uint64_t now = millis();
        if((now - heartbeatTimestamp) > HEARTBEAT_INTERVAL) {
            heartbeatTimestamp = now;
            // socket.io heartbeat message
            webSocket.sendTXT("2");
        }
    }`
As far as I can see, this is not possible with the current implementation. Although I am grateful for the new capability of event handling, please add acknowledgement functionality or the connection will die after 30 seconds.

**Comment by Links2004** [comment]
the heartbeat is not acknowledgement and heartbeat is Implemented and done automatically.
you can take a look at:

https://github.com/Links2004/arduinoWebSockets/blob/master/src/SocketIOclient.cpp#L94-L98

Note: eIOtype_PING == "2" 

and answering to PING requests is here:
https://github.com/Links2004/arduinoWebSockets/blob/master/src/SocketIOclient.cpp#L121-L125

the acknowledgement is a other level of the protocol and has nothing to do with keeping it alive.
acknowledgement are mostly used to trigger callbacks for messages on the server side.

@anselmpaul if your connection gets disconnected, please open a new ticket with debug logging enabled then we can see where the problem is located.

@Sanuja91 
`sIOtype_ACK` will be done soon, may be end of this week.

**Comment by Sanuja91** [comment]
@Links2004 I worked out a solution for the acknowledgements from the server side by using an event emitter with a custom event ID which is sent with the initial socket message from the server... Once the device gets the message it sends an acknowledgment message through socketIO which then triggers an event on the server.. If the event is triggered before a timeout on the server, the server knows that the client has received the message.. It works flawlessly! 

Drop a message once sIOtype_ACK is done =)

**Comment by Links2004** [comment]
@Sanuja91 
latest master has support for it.

https://github.com/Links2004/arduinoWebSockets/blob/master/examples/esp8266/WebSocketClientSocketIOack/WebSocketClientSocketIOack.ino
